src/inference/dreamEubo.ad.js

```diff
@@ -0,0 +1,189 @@
+
+      var dream = 0;
+      var grad = {};
+
+      return util.cpsForEach(
+
+          // Body.
+          function(record, i, records, next) {
+            return this.estimateGradient(record, function(g, dream_i) {
+              paramStruct.addEq(grad, g); // Accumulates gradient estimates.
+              dream += dream_i;
+              return next();
+            });
+          }.bind(this),
+
+          // Continuation.
+          function() {
+            paramStruct.divEq(grad, this.records.length);
+            dream /= this.records.length; // Averages estimations.
+            env.coroutine = this.coroutine;
+            return this.cont(grad, dream);
+          }.bind(this),
+
+          this.records);
+
+    },
+
+    // Computes a single sample estimate of the gradient,
+    // which in case of EUBO is the minus guide scoring on
+    // the target recorded samples.
+    estimateGradient: function(record, cont) {
+
+      // Makes record available throughout the module.
+      this.currRecord = record;
+
+      // paramsSeen tracks the AD nodes of all parameters seen during
+      // a single execution. These are the parameters for which
+      // gradients will be computed.
+      this.paramsSeen = {};
+      this.logq = 0;
+
+      return this.wpplFn(_.clone(this.s), function() {
+
+        var objective = -this.logq;
+
+        if (ad.isLifted(objective)) { // Handles programs with zero random choices.
+          objective.backprop();
+        }
+
+        var grads = _.mapObject(this.paramsSeen, function(params) {
+          return params.map(ad.derivative);
+        });
+
+        var logp = ad.value(this.currRecord.trace.score);
+        var logq = ad.value(this.logq);
+        return cont(grads, logp - logq);
+        
+      }.bind(this), this.a);
+
+    },
+
+    // Since we already have all the records from a previous pass,
+    // sample retrieves a previous sampled value from records for the
+    // corresponding variable in the program that is asked to be sampled.
+    // Inside mapData, it also accumulates the guide score into the objective.
+    sample: function(s, k, a, dist, options) {
+      // If guide distribution is not provided then we use mean field
+      var guideDist = (options && options.guide) || guide.independent(dist, a, env);
+      var rel = util.relativizeAddress(env, a);
+      var guideVal = this.currRecord.trace.findChoice(
+          this.currRecord.trace.baseAddress + rel).val;
+      assert.notStrictEqual(guideVal, undefined);
+
+      if (isInsideMapData()) {
+
+        // We unlift guideVal to maintain the separation between the ad
+        // graph we're building in order to optimize the parameters and
+        // any ad graphs associated with the example traces. (The
+        // choices in an example trace can be ad nodes when they are
+        // generated with SMC + HMC rejuv.)
+        var _guideVal = ad.value(guideVal);
+        
+        var _guideScore = guideDist.score(_guideVal);
```

Comment:
I should explain why you're seeing a bunch of variable names prefixed with underscores. When a piece of code has to deal with both a node in the ad graph and the node's value, I often use similar variables for both but prefix the name of the latter with an underscore. That's what's happening here with `guideVal` and `_guideVal`. You don't need to do this for the score variable for example.